### ISSUE TITLE
Issue with reading mp3 file

### ISSUE THREAD

**Issue by Nekromancer** [naudio/NAudio#292]
I have huge number of monstrually downsamples mp3 files. 30 seconds each, 11khz, 16 kbps, 16 bits, mono, 60 KB of size. I know a lot of them have different sample rates in frames. But practically all players i checked workes fine and plays this monstrocity. Unfortunetely NAudio cannot deal with it. 

Code:
```
var reader = new Mp3FileReader(filePath);
```
returns an exception:
```
System.InvalidOperationException: Got a frame at sample rate 22050, in an MP3 with sample rate 44100. Mp3FileReader does not support sample rate changes.
   w NAudio.Wave.Mp3FileReader.ValidateFrameFormat(Mp3Frame frame)
   w NAudio.Wave.Mp3FileReader.CreateTableOfContents()
   w NAudio.Wave.Mp3FileReader..ctor(Stream inputStream, FrameDecompressorBuilder frameDecompressorBuilder, Boolean ownInputStream)
   w NAudio.Wave.Mp3FileReader..ctor(String mp3FileName)
```
Is there any way to omit the validation phase and read this mp3 frames? All i want to do with this files is to read Id3v2Tag RawData and then read all their frames like that:
```
Mp3Frame frame;

while ((frame = reader.ReadNextFrame()) != null)
{
     stream.Write(frame.RawData, 0, frame.RawData.Length);
 }
```

**Comment by markheath** [comment]
Have you tried just using `MediaFoundationReader` instead of `Mp3FileReader`? It uses Windows' MP3 support so is generally more reliable.

**Comment by Nekromancer** [comment]
I'm trying to recreate my solution using `MediaFoundationReader`, but i'm not sure how to implement same thing.

My actual code:
```
private static void AddFileToStream(MemoryStream stream, string filePath)
        {
            var reader = new Mp3FileReader(filePath);

            if (stream.Position == 0)
            {
                if (reader.Id3v2Tag == null)
                {
                    var tag = Id3v2Tag.Create(new List<KeyValuePair<string, string>>() { new KeyValuePair<string, string>("year", DateTime.Now.Year.ToString()) });
                    stream.Write(tag.RawData, 0, tag.RawData.Length);
                }
                else
                    stream.Write(reader.Id3v2Tag.RawData, 0, reader.Id3v2Tag.RawData.Length);
            }

            Mp3Frame frame;

            while ((frame = reader.ReadNextFrame()) != null)
            {
                stream.Write(frame.RawData, 0, frame.RawData.Length);
            }

            reader.Close();
        }
```
It's a function that reads whole mp3 file and write it's frames to `MemoryStream`. The idea is to join a few mp3 files into single mp3 file. It doesn't have to be perfectly aligned, but it have to be pretty small (buffer of this `MemoryStream` is sent to website front-end with mp3 mime type).

This function might be called as that:
```
var files = Directory.GetFiles(path, "*.mp3");

            var output = new MemoryStream();
            var fileStream = new FileStream("output.mp3", FileMode.Create);

            try
            {
                foreach (var file in files)
                {
                    AddFileToStream(output, file);
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
            }
            output.WriteTo(fileStream);
            fileStream.Close();
            output.Close();
```
It's simplified code to show the problem.

**Comment by markheath** [comment]
OK, you can't append MP3 files using that technique. You'd need to concatenate the 3 files and then re-encode them to MP3 with something like LAME.

**Comment by Nekromancer** [comment]
Oh well.. I think doing that is too much overhead for this project. 

Method i'm using now, with `Mp3FileReader`, is working fine with most of audio files. But unfortunetely not with all of them. Is this bug fixable?

**Comment by markheath** [comment]
Yes, I'm sure its fixable. The trouble with MP3 is that there are a lot of different types of them out there and sometimes they genuinely are not well-formed. I generally recommend using MediaFoundationReader where possible for that reason as its likely to cope with a much broader range of MP3 files than the NAudio code.

**Comment by frankpost** [comment]
Hello Mark ... regarding the Mp3FileReader issue and your recommendation to use the 
MediaFoundationReader currently my application takes in custom encrypted mp3 files and decrypts them into a MemoryStream.  This Stream is then passed into the Mp3FileReader.  Unfortunately I am receiving too often the InvalidOperationException "Got a frame at sample rate N, in an MP3 with sample rate X. Mp3FileReader does not support sample rate changes."  I would like to use the 
MediaFoundationReader however, the constructor only takes a file path and in the case of my application I do not want to expose non-encrypted mp3 audio files.  I have researched how/if the  MediaFoundationReader can be modified to use a stream ( IMFByteStream ) but there does not seem to be viable solutions to this.  Are you aware of how this can be accomplished?

Thank you for your time Mark.

**Comment by vencious** [comment]
Hi, i get the same issues as described in the first post and i am actually using the plugin in Unity engine. In general it works fine except the problem that it can not load certain MP3 files which i don't know why is really !

However, i also tried to use "MediaFoundationReader" as suggested above but the engine just crashes at this part of the code :

`mediaReader= new MediaFoundationReader(musicTrackPath);` - Crashes the game engine

`mp3Reader = new Mp3FileReader(musicTrackPath);` - Works for most MP3 files but not all !

I simply changed the variable type from "Mp3FileReader" to "MediaFoundationReader".
Is there any way i can deal with this !?

Any help is really much appreciated ! Thanks !

**Comment by markheath** [comment]
It depends what version of Windows you are on as to whether MediaFoundationReader will work or not. It uses a different set of Windows APIs and codecs to Mp3FileReader (which uses ACM codecs by default)

**Comment by vencious** [comment]
> It depends what version of Windows you are on as to whether MediaFoundationReader will work or not. It uses a different set of Windows APIs and codecs to Mp3FileReader (which uses ACM codecs by default)

MediaFoundationReader crashes on latest Windows 10 x64 ! I am now using "Mp3FileReader" for mp3 files and try catch which mp3 files can not play and try to play another one. It seems to work for now as a workaround.

May i ask which are recommended ( or min ) Windows versions for Mp3FileReader, WaveFileReader  and MediaFoundationReader ?

**Comment by markheath** [comment]
What error does MediaFoundationReader crash with?
Desktop versions of Windows tend to come with more codecs pre-installed than server versions of Windows. I'd expect most MP3s to be playable with either Mp3FileReader or MediaFoundationReader on desktop editions of Windows 10 out of the box.

**Comment by markheath** [comment]
Closing this as we have fixed these issues with false MP3 sample rate changes in NAudio 3
You can also now read MP3 files with MediaFoundationReader (recommended on Windows) and SoundFileReader (recommended for cross platform)
